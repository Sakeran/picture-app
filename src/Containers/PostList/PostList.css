.PostList-masonry {
  margin: 0 auto;
}
