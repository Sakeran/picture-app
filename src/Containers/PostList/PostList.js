import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import PostCard from '../../Components/PostCard/PostCard';

import Masonry from 'react-masonry-component';
import './PostList.css';

class PostList extends React.Component {

  componentDidMount() {
    // Check each post id as it is passed in. Query for any data that does
    // not exist in store. The list will only display once everything is
    // finished loading.
    this.props.postIDs.forEach(e => {
      if(!this.props.posts[e]) {
        return this.props.getPost(e)
        .then(post => {
          this.props.addPost(post);
        });
      }
    });
  }

  render() {
    return (
      <Masonry className="PostList-masonry" options={{fitWidth: true, gutter: 6}}>
        {this.props.postIDs.map(e => (
          this.props.posts[e] &&
          <PostCard key={e} post={this.props.posts[e]} />
        ))}
      </Masonry>
    );
  }
}

PostList.propTypes = {
  posts: PropTypes.object.isRequired,
  addPost: PropTypes.func.isRequired,
  getPost: PropTypes.func.isRequired,
  postIDs: PropTypes.array.isRequired
};

const mapStateToProps = (state) => ({
  posts: state.posts
});

const mapDispatchToProps = (dispatch) => ({
  addPost: (post) => dispatch({type:'ADD_POST', post: post})
});

export { PostList }
export default connect(mapStateToProps, mapDispatchToProps)(PostList);
