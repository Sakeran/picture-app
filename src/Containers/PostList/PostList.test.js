import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { PostList } from './PostList';

// Mocks/Factories

const makePost = (id) => ({
  id: id,
  type: 'image',
  image: 'http://localhost:3001/exampleImg.png',
  title: 'test card ' + id,
  youtubeID: null
});

const getPost = (id) => Promise.resolve(makePost(id));


test('renders PostList element correctly', () => {
  const ids = ['11111','22222','33333','44444','55555'];
  const posts = {};
  ids.forEach(e => { posts[e] = makePost(e)});
  const component = shallow(<PostList postIDs={ids} getPost={getPost} posts={posts}/>);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('queries for posts not present in store', (done) => {
  const ids = ['11111','22222','33333','44444','55555'];
  const posts = {};
  const addPost = jest.fn();
  const component = mount(<PostList postIDs={ids} getPost={getPost} posts={posts} addPost={addPost}/>);
  process.nextTick(() => {
    expect(addPost.mock.calls).toHaveLength(5);
    done();
  });
});

test('does not query for posts present in store', (done) => {
  const ids = ['11111','22222','33333','44444','55555'];
  const posts = {};
  ids.forEach(e => { posts[e] = makePost(e)});
  const addPost = jest.fn();
  const component = mount(<PostList postIDs={ids} getPost={getPost} posts={posts} addPost={addPost}/>);
  process.nextTick(() => {
    expect(addPost.mock.calls).toHaveLength(0);
    done();
  });
});
