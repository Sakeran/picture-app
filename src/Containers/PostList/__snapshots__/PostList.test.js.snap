// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PostList element correctly 1`] = `
<MasonryComponent
  className="PostList-masonry"
  disableImagesLoaded={false}
  elementType="div"
  enableResizableChildren={false}
  onLayoutComplete={[Function]}
  onRemoveComplete={[Function]}
  options={
    Object {
      "fitWidth": true,
      "gutter": 6,
    }
  }
  updateOnEachImageLoad={false}
>
  <PostCard
    post={
      Object {
        "id": "11111",
        "image": "http://localhost:3001/exampleImg.png",
        "title": "test card 11111",
        "type": "image",
        "youtubeID": null,
      }
    }
  />
  <PostCard
    post={
      Object {
        "id": "22222",
        "image": "http://localhost:3001/exampleImg.png",
        "title": "test card 22222",
        "type": "image",
        "youtubeID": null,
      }
    }
  />
  <PostCard
    post={
      Object {
        "id": "33333",
        "image": "http://localhost:3001/exampleImg.png",
        "title": "test card 33333",
        "type": "image",
        "youtubeID": null,
      }
    }
  />
  <PostCard
    post={
      Object {
        "id": "44444",
        "image": "http://localhost:3001/exampleImg.png",
        "title": "test card 44444",
        "type": "image",
        "youtubeID": null,
      }
    }
  />
  <PostCard
    post={
      Object {
        "id": "55555",
        "image": "http://localhost:3001/exampleImg.png",
        "title": "test card 55555",
        "type": "image",
        "youtubeID": null,
      }
    }
  />
</MasonryComponent>
`;
